docs/development_guidelines.md

```diff
@@ -76,3 +76,60 @@ but also helps fellow contributors making adjustments to your code in the future
 By default, Home Assistant will statically check for type hints in our automated CI process.
 Python modules can be included for strict checking, if they are fully typed, by adding an entry
 to the `.strict-typing` file in the root of the Home Assistant Core project.
+
+### Assertions and error checking
+
+Use `assert` statements only for type annotations to help mypy, and for pytest assertions. This is because assertions can be turned off so they cannot be relied upon for runtime checks. To check for errors, use `if` and log then `return`, or `raise` an exception as appropriate.
+
+This is a good assertion, used to help type checking:
+
+```python
+class Foo:
+    optional_var: Something | None
+
+    def bar() -> None
+        if self.optional_var is None:
+            self._do_thing_without_optional_var()
+        else:
+            self._do_thing_with_optional_var()
+
+    def _do_thing_with_optional_var() -> None:
+        # This will only be called when optional_var is set, but mypy doesn't know that
+        assert self.optional_var is not None
```

Comment:
The point here is, this should not be asserted. Instead, it should raise. If this method is called when there is not an option available, that is unexpected behavior.

Typing in this case, warns you that this needs to be guarded.

Reply:
I'd agree with you if this was part of a public API, but this example is for a private method. Here we know that `optional_var` can never be `None` when `_do_thing_with_optional_var` is called, as it is only called via `bar` (which does the check for `None`). The `assert` codifies the preconditions we know to be true.

Raising an exception would look like:
```python
    def _do_thing_with_optional_var() -> None:
        if self.optional_var is None:
            raise SomeError
```

There are two downsides to this, compared to an assert:
1) An `assert` can be optimised away in production code, but the `if` can't, so there's a performance impact.
2) We will never hit the `raise SomeError` line by calling the public API, so if we're following the HA test guidelines then this line of code will never be tested. It's dead code that could hide bugs, and it'll drop the coverage rate.

Reply:
> but this example is for a private method.

It is not, it is marked protected (single underscore). Thus consumable by child classes.

Reply:
Ok, it's a protected member by the principle of "[we are all responsible users](https://docs.python-guide.org/writing/style/#we-are-all-responsible-users)". If the author of a child class is calling an underscore-prefixed method then they should know what they're doing, and that includes satisfying all preconditions. 

The assert still works for debugging the case where a child or client of the class has not respected the method preconditions, and will pick it up during test runs.

I don't think it's worthwhile defending against the case of someone slipping an invalid value through in production that was missed during testing, debugging, and code reviews. This can lead to an impossible situation of trying to hit 100% code coverage ([as required for Gold quality](https://developers.home-assistant.io/docs/integration_quality_scale_index/#gold-) on config flow) while also not accessing protected members from tests ([as required by PyLint](http://pylint-messages.wikidot.com/messages:w0212)).

Reply:
🤷

Almost all assertions we do, are unneeded or caused by improper design (or could be fixed by using other designs. For example, make better use of injecting dependencies, instead of locally spawning them). I agree, there are very very rare cases that could allow for it and be a better solution.

I feel like adding this "guideline" adds a free card to use them. IMHO we should document they are NOT normal to use and thus only allow them in exceptional cases.

The approach in this guideline advertises "good use cases", however, in most cases it's not that simple, and even in this example given here, lacking the context of the bigger design and picture.

Therefore; I still think we should strongly discourage their use and actively encourage better design patterns in general.

Besides on a Monopoly game board, I don't think we should ever advertise get out of jail free cards.

If people want to overrule this and still add "asserts" to production code as something that is needed or good; that is fine. I don't support it. Sorry.

Reply:
> :shrug:

Don't know or don't care?

> Almost all assertions we do, are unneeded or caused by improper design (or could be fixed by using other designs. For example, make better use of injecting dependencies, instead of locally spawning them). I agree, there are very very rare cases that could allow for it and be a better solution.
> 
> I feel like adding this "guideline" adds a free card to use them. IMHO we should document they are NOT normal to use and thus only allow them in exceptional cases (no other was out is possible in a sane way).

I've rewritten this PR to describe best-practice for variable/state/return-value validation. Now asserts are covered as an addendum of the exceptional cases.

> The approach in this guideline advertises "good use cases", however, in most cases it's not that simple, and even in this example given here, lacking the context of the bigger design and picture.

I've added context and admonitions to do better.

> Therefore; I still think we should strongly discourage their use and actively encourage better design patterns in general.
> 
> Besides on a Monopoly game board, I don't think we should ever advertise get out of jail free cards.
> 
> If people want to overrule this and still add "asserts" to production code as something that is needed or good; that is fine. I don't support it. Sorry.

Ignoring `assert`in the developer guidelines only adds to confusion for code authors and reviewers. Without a reference for what is accepted or not, you'll get wildly inconsistent code. And then it makes reviewing a painful experience, because the author and reviewer have differing views on what is acceptable. That's why I wrote this PR in the first place.

Reply:
I think we're going into too great detail about the Python language in the PR. I think it's enough to add a short sentence about the `assert` statement:

> Avoid using the `assert` statement in production code, unless it's required for typing purposes and there aren't better alternatives available.

This still leaves room for a case by case judgement and we don't try to describe general Python practices in detail which we shouldn't need to do and also will have a hard time keeping up to date as Python and this project evolves.

Reply:
I've dropped the `Assert` subsection down to the single suggested sentence. I'm not super happy with leaving it this vague, as it now feels like whether `assert` will pass code review will depend on the personal preferences of the reviewer.

E.g. What about assertion for debugging function pre- and post-conditions? How far do we have to change the code to avoid `assert` for static typing? What are better alternatives?

But it looks like @balloob and @frenck are in agreement, so I guess this'll do for now.